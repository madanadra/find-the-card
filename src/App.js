import './style/all.css';
import { useEffect, useState } from "react";
import { useCookies } from "react-cookie";
import Menu from "./component/menu";
import Main from "./component/main";
import GameOver from "./component/game-over";
import Level from "./component/level";
import HighScore from "./component/high-score";
import Guide from "./component/guide";
import backgroundAudio from "./asset/sound/bg.mp3";
import musicAudio from "./asset/sound/music.wav";
import clickAudio from "./asset/sound/click.wav";
import winAudio from "./asset/sound/win.wav";
import gameOverAudio from "./asset/sound/gameover.wav";

function App() {
  const [cookies, setCookie] = useCookies(['highScore'])
  const [menu, setMenu] = useState('menu')
  const [number, setNumber] = useState()
  const [open, setOpen] = useState(false)
  const lv = 
  [
    {lv1: [{id: '0'}, {id: '1'}]}, 
    {lv2: [{id: '0'}, {id: '1'}, {id: '2'}]}, 
    {lv3: [{id: '0'}, {id: '1'}, {id: '2'}, {id: '3'}]},
    {lv4: [{id: '0'}, {id: '1'}, {id: '2'}, {id: '3'}, {id: '4'}]},
    {lv5: [{id: '0'}, {id: '1'}, {id: '2'}, {id: '3'}, {id: '4'}, {id: '5'}]}
  ]
  const [cards, setCards] = useState(lv[0].lv1)
  const [highScore, setHighScore] = useState([])
  const [over, setOver] = useState(false)
  const [score, setScore] = useState(0)
  const [bg, setBg] = useState()
  const click = new Audio(clickAudio)
  const gameover = new Audio(gameOverAudio)
  const win = new Audio(winAudio)

  useEffect(() => {
    setBg(new Audio(backgroundAudio));
    if (cookies.highScore) {
      setHighScore(cookies.highScore);
    }
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (bg) {
      bg.loop = true;
      bg.play();
    }
  }, [bg]);

  useEffect(() => {
    setCookie('highScore', JSON.stringify(highScore), { path: '/', expires: new Date(Date.now()+2592000) });
  }, [highScore]); // eslint-disable-line react-hooks/exhaustive-deps

  const playGames = () => {
    setNumber(Math.floor(Math.random() * cards.length).toString());
    setOpen(false);
  }

  const newGames = () => {
    setMenu('play');
    setScore(0);
    playGames();
    bg.pause();
    setBg(new Audio(musicAudio));
  }

  const openCard = (prop) => {
    if (!open) {
      setOpen(true)
      setTimeout(()=>{
        if (prop === number) {
          playGames();
          win.play();
          setScore(score+(100*(cards.length-1)));
        } else {
          setOver(true);
          bg.pause();
          gameover.play();
          setHighScore([...highScore, {score: score, date: Date.now()}].sort((a,b) => b.score-a.score).slice(0,5));
        }
        return () => {};
      }, 1000)
    }
  }

  const GameMenu = () => {
    if (menu === 'menu') {
      return (
        <Menu newGames={newGames} setMenu={setMenu} cards={cards} click={click} />
      );
    } else if (menu === 'play') {
      return (
        <Main score={score} number={number} cards={cards} open={open} openCard={openCard} bg={bg} setBg={setBg} setMenu={setMenu} click={click} />
      );
    } else if (menu === 'level') {
      return (
        <Level cards={cards} setCards={setCards} setMenu={setMenu} lv={lv} click={click} />
      );
    } else if (menu === 'highscore') {
      return (
        <HighScore highScore={highScore} setMenu={setMenu} click={click} />
      );
    } else if (menu === 'guide') {
      return (
        <Guide setMenu={setMenu} click={click} />
      );
    }
  }

  return (
    <div className="container">
      <GameMenu />
      {over && <GameOver newGames={newGames} setMenu={setMenu} setOver={setOver} score={score} setBg={setBg} click={click} />}
    </div>
  );
}

export default App;
