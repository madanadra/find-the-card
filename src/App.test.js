import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import App from './App';

describe('App component', () => {
  afterEach(cleanup);
  test('first', () => {
    const { asFragment } = render(<App />)
    expect(asFragment(<App />)).toMatchSnapshot()
  });

  test('menu', () => {
    render(<App />)
    const playElement = screen.getByText('Play');
    const levelElement = screen.getByText('Level 1');
    const highscoreElement = screen.getByText('High Scores');
    const guideElement = screen.getByText('Guides');
    expect(playElement).toBeInTheDocument();
    expect(levelElement).toBeInTheDocument();
    expect(highscoreElement).toBeInTheDocument();
    expect(guideElement).toBeInTheDocument();
  });

  test('play', () => {
    render(<App />)
    fireEvent.click(screen.getByText('Play'));
    expect(screen.getByText('Back')).toBeInTheDocument();
    expect(screen.getByText('Score: 0')).toBeInTheDocument();
  });

  test('level', () => {
    render(<App />)
    fireEvent.click(screen.getByText('Level 1'));
    expect(screen.getByText('Back')).toBeInTheDocument();
    expect(screen.getByText('Level 1')).toBeInTheDocument();
    expect(screen.getByText('Level 2')).toBeInTheDocument();
    expect(screen.getByText('Level 3')).toBeInTheDocument();
    expect(screen.getByText('Level 4')).toBeInTheDocument();
    expect(screen.getByText('Level 5')).toBeInTheDocument();
  });

  test('highScore', () => {
    render(<App />)
    fireEvent.click(screen.getByText('High Scores'));
    expect(screen.getByText('Back')).toBeInTheDocument();
    expect(screen.getByText('#')).toBeInTheDocument();
    expect(screen.getByText('Score')).toBeInTheDocument();
    expect(screen.getByText('Date')).toBeInTheDocument();
  });

  test('guide', () => {
    render(<App />)
    fireEvent.click(screen.getByText('Guides'));
    expect(screen.getByText('Back')).toBeInTheDocument();
    expect(screen.getByText('Find a green card!')).toBeInTheDocument();
    expect(screen.getByText('Lv 1: 2 Cards & 100 Scores')).toBeInTheDocument();
    expect(screen.getByText('Lv 2: 3 Cards & 200 Scores')).toBeInTheDocument();
    expect(screen.getByText('Lv 3: 4 Cards & 300 Scores')).toBeInTheDocument();
    expect(screen.getByText('Lv 4: 5 Cards & 400 Scores')).toBeInTheDocument();
    expect(screen.getByText('Lv 5: 6 Cards & 500 Scores')).toBeInTheDocument();
  });
});
