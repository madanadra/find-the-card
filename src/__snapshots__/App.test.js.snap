// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component first 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="menu"
    >
      <h1>
        Play
      </h1>
      <h1>
        Level 1
      </h1>
      <h1>
        High Scores
      </h1>
      <h1>
        Guides
      </h1>
    </div>
  </div>
</DocumentFragment>
`;
